waspc/test/AnalyzerTest.hs

```diff
@@ -189,6 +192,74 @@ spec_Analyzer = do
               ]
       isRight (analyze source) `shouldBe` True
 
+    describe "Returns correct error message" $ do
+      let errorMessageShouldBe analyzeResult (c, msg) = case analyzeResult of
+            Right _ -> error "Test failed: expected AnalyzerError."
+            Left e -> getErrorMessageAndCtx e `shouldBe` (msg, c)
+
+      it "For nested unexpected type error" $ do
+        let source =
+              unlines
+                [ "app MyApp {",
+                  "  title: \"My app\",",
+                  "  dependencies: [",
+                  "    { name: \"bar\", version: 13 },",
+                  "    { name: \"foo\", version: 14 }",
+                  "  ]",
+                  "}"
+                ]
+        analyze source
+          `errorMessageShouldBe` ( ctx 4 29,
+                                   intercalate
+                                     "\n"
+                                     [ "Type error:",
+                                       "  For dictionary field 'dependencies':",
+                                       "    For list element:",
+                                       "      For dictionary field 'version':",
+                                       "        Expected type: string",
+                                       "        Actual type:   number"
+                                     ]
+                                 )
+
+      it "For nested unification type error" $ do
+        let source =
+              unlines
+                [ "app MyApp {",
+                  "  title: \"My app\",",
+                  "  dependencies: [",
+                  "    { name: \"bar\", version: 13 },",
+                  "    { name: \"foo\", version: \"1.2.3\" }",
+                  "  ]",
+                  "}"
+                ]
+        analyze source
+          `errorMessageShouldBe` ( ctx 5 29,
+                                   intercalate
+                                     "\n"
+                                     [ "Type error:",
+                                       "  For dictionary field 'version':",
+                                       "    Can't mix following types:",
```

Comment:
this is an awesome error message, super human-readable and easy to understand, with no parsing lingo

Reply:
@matijaSos I do like the idea of putting the proximal error first, and then expounding on that with location details below. Good idea 👍🏻 And it does mirror how Haskel does it... I just forced a random error:

```haskell
 • Could not deduce (WriteableMonad
                          ((->) GHC.IO.Exception.IOErrorType))
        arising from a use of ‘throwIO’
      from the context: WriteableMonad m
        bound by the type signature for:
                   write :: forall (m :: * -> *).
                            WriteableMonad m =>
                            Path' Abs (SP.Dir ProjectRootDir) -> CopyFileDraft -> m ()
        at src/Wasp/Generator/FileDraft/CopyFileDraft.hs:31:3-7
    • In the second argument of ‘when’, namely
        ‘(throwIO
            mkIOError doesNotExistErrorType
            "Source file of CopyFileDraft does not exist." Nothing
            (Just srcFilePath))’
      In the expression:
        when
          (_failIfSrcDoesNotExist draft)
          (throwIO
             mkIOError doesNotExistErrorType
             "Source file of CopyFileDraft does not exist." Nothing
             (Just srcFilePath))
      In a stmt of a 'do' block:
        if srcFileExists then
            do createDirectoryIfMissing
                 True (SP.fromAbsDir $ SP.parent absDraftDstPath)
               copyFile srcFilePath (SP.fromAbsFile absDraftDstPath)
        else
            when
              (_failIfSrcDoesNotExist draft)
              (throwIO
                 mkIOError doesNotExistErrorType
                 "Source file of CopyFileDraft does not exist." Nothing
                 (Just srcFilePath))
   |
40 |           ( throwIO
   |             ^^^^^^^...

```

Reply:
@shayneczyzewski this is great! Haskell errors are notorious but this part it seems they got right :D.